asf_smart_meter_exploration/utils/plotting_utils.py

```diff
@@ -0,0 +1,229 @@
+import matplotlib.patheffects as pe
+import matplotlib.ticker as mtick
+import datetime
+
+from asf_smart_meter_exploration import config, PROJECT_DIR
+
+cluster_plot_folder_path = PROJECT_DIR / config["cluster_plot_folder_path"]
+plot_suffix = config["plot_suffix"]
+
+
+def set_plot_properties(chart):
+    """Set custom size, font and colour properties for an Altair chart.
+
+    Args:
+        chart (alt.Chart): Any Altair chart.
+
+    Returns:
+        alt.Chart: Chart with adjusted properties.
+    """
+    return (
+        chart.properties(width=800, height=300)
+        .configure_axis(labelFontSize=20, titleFontSize=20)
+        .configure_legend(titleFontSize=18, labelFontSize=15)
+        .configure_range(category={"scheme": "dark2"})
+    )
+
+
+def plot_observations_and_clusters(
+    data,
+    clusters,
+    filename_infix,
+    normalised=True,
+    ylabel="Electricity usage (normalised)",
+    ymin=0,
+    ymax=0.2,
+):
+    """Produce and save plot featuring all household usage lines and thicker lines denoting clusters.
+
+    Args:
+        data (pd.DataFrame): Household usage data.
+        clusters (list): Cluster designations.
+        filename_infix (str): Description of variant (e.g. "normalised_usage").
+            Appears in filename and plot title.
+        normalised (bool, optional): Whether the measure has been normalised
+            (determines whether or not to display y axis as a percentage).
+            Defaults to True.
+        ylabel (str, optional): y axis label. Defaults to "Electricity usage (normalised)".
+        ymin (int, optional): Minimum value on y axis. Defaults to 0.
+        ymax (float, optional): Maximum value on y axis. Defaults to 0.2.
+    """
+    # Using matplotlib here due to issues with how Altair deals with times
+    # Line below makes times work
+    pd.plotting.register_matplotlib_converters()
+
+    fig, ax = plt.subplots()
+
+    fig.set_size_inches(12, 6)
+
+    # Plot individual lines with high transparency
+    for i in range(len(clusters)):
+        ax.plot(data.iloc[i], alpha=0.05, color=f"C{clusters[i]}")
+
+    # Plot cluster lines, thicker and opaque with borders
+    for i in range(max(clusters) + 1):
+        cluster_average = data.loc[clusters == i].mean(numeric_only=True)
+        ax.plot(
+            cluster_average,
+            lw=3,
+            path_effects=[
+                pe.Stroke(linewidth=3, foreground="black"),
+                pe.Stroke(linewidth=2, foreground=f"C{i}"),
+            ],
+            label=i,
+        )
+
+    ax.legend(title="Cluster", loc="upper right")
+    ax.set_xlabel("Time")
+    ax.set_ylabel(ylabel)
+    ax.set_xlim(datetime.time(0, 0, 0), datetime.time(23, 30, 00))
+    ax.set_ylim(ymin, ymax)
+    if normalised:
+        ax.yaxis.set_major_formatter(mtick.PercentFormatter(1, decimals=0))
+
+    ax.set_xticks(["00:00:00", "06:00:00", "12:00:00", "18:00:00"])
+    ax.locator_params(axis="y", nbins=4)
```

Comment:
Weird! Possibly because with certain ymin/ymax values, 4 gives awkward numbers so it uses something close to 4 instead? I'm not sure this line is even really needed as the default bins are fine, so will remove